kaspresso/src/main/kotlin/com/kaspersky/kaspresso/device/bluetooth/BluetoothImpl.kt

```diff
@@ -0,0 +1,120 @@
+    private val logger: UiTestLogger,
+    private val targetContext: Context,
+    private val adbServer: AdbServer
+) : Bluetooth {
+
+    companion object {
+        private const val CMD_STATE_ENABLE = "enable"
+        private const val CMD_STATE_DISABLE = "disable"
+        private const val BLUETOOTH_STATE_CHANGE_CMD = "svc bluetooth"
+        private const val BLUETOOTH_STATE_CHANGE_ROOT_CMD = "su 0 svc bluetooth"
+        private const val BLUETOOTH_STATE_CHECK_CMD = "settings get global bluetooth_on"
+        private const val BLUETOOTH_STATE_CHECK_RESULT_ENABLED = "1"
+        private const val BLUETOOTH_STATE_CHECK_RESULT_DISABLED = "0"
+        private val ADB_RESULT_REGEX = Regex("exitCode=(\\d+), message=(.+)")
+    }
+
+    private val flakySafetyAlgorithm = FlakySafetyAlgorithm(logger)
+    private val flakySafetyParams: FlakySafetyParams
+        get() = FlakySafetyParams(
+            timeoutMs = 1000,
+            intervalMs = 100,
+            allowedExceptions = setOf(AdbServerException::class.java)
+        )
+
+    override fun enable() {
+        toggleBluetooth(enable = true)
+    }
+
+    override fun disable() {
+        toggleBluetooth(enable = false)
+    }
+
+    /**
+     * Toggles Bluetooth state
+     * Tries, first and foremost, to send ADB command. If this attempt fails,
+     * opens Android Settings screen and tries to switch Bluetooth setting thumb.
+     */
+    private fun toggleBluetooth(enable: Boolean) {
+        if (isBluetoothNotSupported()) {
+            logger.i("Bluetooth is not supported")
+            return
+        }
+        logger.i("${if (enable) "En" else "Dis"}able bluetooth")
+        if (!changeBluetoothStateUsingAdbServer(enable, BLUETOOTH_STATE_CHANGE_ROOT_CMD) &&
+            !changeBluetoothStateUsingAdbServer(enable, BLUETOOTH_STATE_CHANGE_CMD)
+        ) {
+             toggleBluetoothUsingAndroidSettings(enable)
+             logger.i("Bluetooth ${if (enable) "en" else "dis"}abled")
+        }
+    }
+
+    /**
+     * Tries to change Bluetooth state using AdbServer if it is available
+     * @return true if Bluetooth state changed or false otherwise
+     */
+    private fun changeBluetoothStateUsingAdbServer(isEnabled: Boolean, changeCommand: String): Boolean =
+        try {
+            val (state, expectedResult) = when (isEnabled) {
+                true -> CMD_STATE_ENABLE to BLUETOOTH_STATE_CHECK_RESULT_ENABLED
+                false -> CMD_STATE_DISABLE to BLUETOOTH_STATE_CHECK_RESULT_DISABLED
+            }
+            adbServer.performShell("$changeCommand $state")
+            flakySafetyAlgorithm.invokeFlakySafely(flakySafetyParams) {
+                val result = adbServer.performShell(BLUETOOTH_STATE_CHECK_CMD)
+                if (parseAdbResponse(result)?.trim() == expectedResult) true else
+                    throw AdbServerException("Failed to change Bluetooth state using ABD")
+            }
+        } catch (e: AdbServerException) {
+            false
```

Comment:
Catching AdbServerException here may lead to the missing the fact that adb server is down or broken. Please, check that test crashes when adb server is disabled. I would throw exception in the toggleBluetooth method if bluetooth state didn't change since we failed to set the test in the desired state

Reply:
Understandable